azure-cs-aks-private-container-registry/Program.cs

```diff
@@ -0,0 +1,216 @@
+				ServicePrincipalId = adSp.Id,
+				Value = randomPassword,
+				EndDate = "2099-01-01T00:00:00Z",
+			});
+
+			// Grant networking permissions to the SP (needed e.g. to provision Load Balancers).
+			var assignment = new Assignment("role-assignment", new AssignmentArgs {
+				PrincipalId = adSp.Id,
+				Scope = resourceGroup.Id,
+				RoleDefinitionName = "Network Contributor"
+			});
+
+			// Create a Virtual Network for the cluster.
+			var vnet = new VirtualNetwork("vnet", new VirtualNetworkArgs {
+				ResourceGroupName = resourceGroup.Name,
+				AddressSpaces = { "10.2.0.0/16" },
+			});
+
+			// Create a Subnet for the cluster.
+			var subnet = new Subnet("subnet", new SubnetArgs {
+				ResourceGroupName = resourceGroup.Name,
+				VirtualNetworkName = vnet.Name,
+				AddressPrefix = "10.2.1.0/24",
+			});
+
+			// Now allocate an AKS cluster.
+			var cluster = new KubernetesCluster("aksCluster", new KubernetesClusterArgs {
+				ResourceGroupName = resourceGroup.Name,
+				AgentPoolProfiles = new KubernetesClusterAgentPoolProfilesArgs {
+					Name = "aksagentpool",
+					Count = 3,
+					VmSize = "Standard_B2s",
+					OsDiskSizeGb = 30,
+					VnetSubnetId = subnet.Id
+				},
+				DnsPrefix = "sampleaks",
+				LinuxProfile = new KubernetesClusterLinuxProfileArgs {
+					AdminUsername = "aksuser",
+					SshKey = new KubernetesClusterLinuxProfileSshKeyArgs {
+						KeyData = sshPublicKey,
+					},
+				},
+				ServicePrincipal = new KubernetesClusterServicePrincipalArgs {
+					ClientId = adApp.ApplicationId,
+					ClientSecret = adSpPassword.Value,
+				},
+				KubernetesVersion = "1.15.7",
+				RoleBasedAccessControl = new KubernetesClusterRoleBasedAccessControlArgs { Enabled = true },
+				NetworkProfile = new KubernetesClusterNetworkProfileArgs {
+					NetworkPlugin = "azure",
+					DnsServiceIp = "10.2.2.254",
+					ServiceCidr = "10.2.2.0/24",
+					DockerBridgeCidr = "172.17.0.1/16",
+				},
+			});
+
+			// Create a k8s provider pointing to the kubeconfig.
+			var k8sProvider = new Pulumi.Kubernetes.Provider("k8s", new Pulumi.Kubernetes.ProviderArgs {
+				KubeConfig = cluster.KubeConfigRaw
+			});
+
+			// Create a Container Registry.
+			var registry = new Registry("acregistry", new RegistryArgs {
+				ResourceGroupName = resourceGroup.Name,
+				Sku = "Basic",
+				AdminEnabled = true
+			});
+
+			// Build & push the ample application to the registry.
+			var applicationName = "sample-application";
+			var imageName = registry.LoginServer.Apply(value => $"{value}/{applicationName}");
+			
+			var image = new Image(applicationName, new ImageArgs {
+				Build = "./SampleApplication",
+				Registry = new ImageRegistry {
+					Server = registry.LoginServer,
+					Username = registry.AdminUsername,
+					Password = registry.AdminPassword
+				},
+				ImageName = imageName
+			}, new ComponentResourceOptions {
+				Provider = k8sProvider
+			});
```

Comment:
I suggest you creating `options` once and then passing the same object everywhere

Reply:
Makes sense. Except that some objects accept `ComponentResourceOptions`, whilst others accept `CustomResourceOptions` (the majority of cases).

I will create a `CustomResourceOptions` object once and pass it along where applicable, and leave the `ComponentResourceOptions` as-is, since it's the only occurrence.